src/layers/logging.rs

```diff
@@ -86,6 +94,140 @@ impl Layer for LoggingLayer {
     }
 }
 
+/// `LoggingReader` is a wrapper of `BytesReader`, with logging functionality.
+pub struct LoggingReader {
+    scheme: Scheme,
+    path: String,
+    op: Operation,
+    have_read: usize,
+    to_read: Option<u64>,
+    inner: BytesReader,
+}
+
+impl LoggingReader {
+    pub(crate) fn make_reader(
+        scheme: Scheme,
+        op: Operation,
+        path: &str,
+        size: Option<u64>,
+        reader: BytesReader,
+    ) -> BytesReader {
```

Comment:
We can return `Self` here, and let users to decide whether to box.

Reply:
I think it's ok to box up since the struct is not exposed to users.

Reply:
> I think it's ok to box up since the struct is not exposed to users.

Ok, your decide.